src/common/Throttle.cc

```diff
@@ -650,3 +650,123 @@ void OrderedThrottle::complete_pending_ops(UNIQUE_LOCK_T(m_lock)& l) {
     ++m_complete_tid;
   }
 }
+
+uint64_t TokenBucketThrottle::Bucket::get(uint64_t c) {
+  if (0 == max) {
+    return 0;
+  }
+
+  uint64_t got = 0;
+  if (remain >= c) {
+    // There is enough token in bucket, take c.
+    got = c;
+    remain -= c;
+  } else {
+    // There is not enough, take all remain.
+    got = remain;
+    remain = 0;
+  }
+  return got;
+}
+
+uint64_t TokenBucketThrottle::Bucket::put(uint64_t c) {
+  if (0 == max) {
+    return 0;
+  }
+  if (c) {
+    // put c tokens into bucket
+    uint64_t current = remain;
+    if ((current + c) <= (uint64_t)max) {
+      remain += c;
+    } else {
+      remain = (uint64_t)max;
+    }
+  }
+  return remain;
+}
+
+void TokenBucketThrottle::Bucket::set_max(uint64_t m) {
+  if (remain > m)
+    remain = m;
+  max = m;
+}
+
+TokenBucketThrottle::TokenBucketThrottle(
+    CephContext *cct,
+    uint64_t capacity,
+    uint64_t avg,
+    SafeTimer *timer,
+    Mutex *timer_lock)
+  : m_cct(cct), m_throttle(m_cct, "token_bucket_throttle", capacity),
+    m_avg(avg), m_timer(timer), m_timer_lock(timer_lock),
+    m_lock("token_bucket_throttle_lock")
+{
+  Mutex::Locker timer_locker(*m_timer_lock);
+  schedule_timer();
+}
+
+TokenBucketThrottle::~TokenBucketThrottle()
+{
+  {
+    Mutex::Locker timer_locker(*m_timer_lock);
+    cancel_timer();
+  }
+  while (!m_blockers.empty()) {
+    Blocker *b = m_blockers.front();
+    m_blockers.pop_front();
+    m_lock.Unlock();
+    b->ctx->complete(0);
+    m_lock.Lock();
+    delete b;
+  }
+}
+
+void TokenBucketThrottle::set_max(uint64_t m) {
+  Mutex::Locker lock(m_lock);
+  m_throttle.set_max(m);
+}
+
+void TokenBucketThrottle::set_average(uint64_t avg) {
+  m_avg = avg;
+}
+
+void TokenBucketThrottle::add_tokens() {
+  // check the m_blockers from head to tail, if blocker can get
+  // enough tokens, let it go.
+  while (!m_blockers.empty()) {
+    Blocker *blocker = m_blockers.front();
+    uint64_t got = m_throttle.get(blocker->tokens_requested);
+    if (got == blocker->tokens_requested) {
+      // got enough tokens for front.
+      m_blockers.pop_front();
+      m_lock.Unlock();
+      m_timer_lock->Unlock();
```

Comment:
It seems less than ideal to keep unlocking and locking two locks here -- and potentially racy if a shutdown occurs when the locks are released. One way to solve this would be to ensure that the destructor cannot fire when this timer action is in-progress (via a boolean and a condition variable) or pass a ```ContextWQ``` to this class so that it can just enqueue the contexts w/o having to drop the lock.

Reply:
I will introduce a m_stopping for this case. But we need to unlocking the two locks here, otherwise there will be a possible cycle locking.
```
Thread 1                                                                       Thread 2
timer_lock.lock()              
m_lock.lock()
timer_lock.unlock()
                                                                                      timer_lock.lock()
                                                                                      m_lock.lock()        <-----there is a dead lock
```

Reply:
Unfortunately, I don't think the ```m_stopping``` helps for all cases. If the throttle is destructed while the locks are released, it will attempt to relock an destroyed mutex on line 751. The destructor would need to wait on a new ```m_callback_in_progress``` variable and wait on a new condition variable.